### ISSUE TITLE
Proposal: Programmatic API for running cucumber-js

### ISSUE THREAD

**Issue by davidjgoss** [cucumber/cucumber-js#1711]
## Problem

Currently we don't have a good way to programmatically run cucumber-js. The need is from two angles:

- Testing - for the CCK and our acceptance tests in the project
- Projects that use cucumber-js as part of a framework (e.g. Serenity, Stryker)

## As-is

What tends to happen at the moment is a new instance of `Cli` is created with [strung-together argv input](https://github.com/cucumber/cucumber-js/blob/main/compatibility/cck_spec.ts#L41). It's obviously very unweildy and also [isn't on the public API]([https://github.com/cucumber/cucumber-js/issues/1710](https://github.com/cucumber/cucumber-js/issues/1710)).

Sometimes (possibly due to perceived fragility of the above) frameworks will just rely on the cucumber-js CLI but struggle to find ways to integrate and have their own options.

The [`Runtime` class](https://github.com/cucumber/cucumber-js/blob/main/src/runtime/index.ts#L42) is currently part of the public API but it's not useful in these contexts, depending on the pickles and support code to be provided by the caller.

## Proposal

Two components in the project:

### `runCucumber`

New async function that executes a test run in-process. Responsibilities:

- Accept an options object with a nice interface
- Do the "work" of resolving pickles, loading support code, running test cases, orchestrating formatters
- Return a promise that resolves to a result

This would be part of the public API and we'd encourage framework maintainers to use it when "wrapping" cucumber-js. We'd also use it for our own testing.

As much as possible, it would avoid direct interaction with `process`, instead accepting normalised options and stream interfaces for output, and leaving it to the caller to decide how to exit based on the result or an unhandled error.

Also `Runtime` should come off the public API as it's really an internal thing.

### CLI

Effectively a "client" of `runCucumber`. Responsibilities:

- Parse input from process arguments (via commander)
- Load profiles and mix those arguments in as needed
- Call `runCucumber` with the resolved options
- Exit as appropriate based on the results

This would continue to *not* be on the public API. Also it would only use functions/interfaces that *are* on the public API, such that we could easily break it out into its own package at some point, as is a common pattern now with projects like [Jest]([https://github.com/facebook/jest/tree/master/packages](https://github.com/facebook/jest/tree/master/packages)).

This decoupling also paves the way for some interesting new CLI features without having them bleed into the internals, e.g.:

- `--gui` for the cucumber-electron stuff
- `--interactive` for quick targeted reruns when TDD'ing

## Timescale

We'd target this at the upcoming 8.0.0 release. I'm ready to gets started on this today.

**Comment by davidjgoss** [comment]
Paging for initial feedback: @aslakhellesoy @charlierudolph @aurelien-reeves @mattwynne @nicojs @jan-molak

**Comment by aslakhellesoy** [comment]
I love this proposal! We already an API like this in fake-cucumber's [runCucumber](https://github.com/cucumber/common/blob/main/fake-cucumber/javascript/src/runCucumber.ts)

**Comment by davidjgoss** [comment]
@aslakhellesoy I may have drawn some inspiration from that...

**Comment by jan-molak** [comment]
@davidjgoss - sounds great! 

For your reference, here's how Serenity/JS invokes Cucumber at the moment - [`CucumberCLIAdapter`](https://github.com/serenity-js/serenity-js/blob/master/packages/cucumber/src/cli/CucumberCLIAdapter.ts)
And here's the logic around converting configuration parameters to argv - [`CucumberOptions`](https://github.com/serenity-js/serenity-js/blob/master/packages/cucumber/src/cli/CucumberOptions.ts).

Being able to provide an options object instead of argv would be much nicer 👍🏻

**Comment by aurelien-reeves** [comment]
Love it!

**Comment by aurelien-reeves** [comment]
While specifying that new public API, we may also consider what recently happened with issue #1489 and think about providing public APIs to have more and better interaction with the filters and the resulting features under test

**Comment by aslakhellesoy** [comment]
And to add to that - let's make something that would make it easier for Cucumber-Electron too:

* https://github.com/cucumber/cucumber-electron/blob/1c2faaa77065754d1ca17dce27fff1c421d226d0/src/renderer/main.js#L49
* https://github.com/cucumber/cucumber-electron/blob/0617ab15545ac5ceb4587599e589fefd0b670142/src/renderer/index.js#L41-L50

**Comment by nicojs** [comment]
Having a public API is better than nothing, so go ahead 👍!

Preferably I would also have an API to load profiles using the same rules as `cucumber-js` does, so I can mimic the exact behavior of a normal cucumber-js call.

StrykerJS will also rely heavily on the [custom_formatters](https://github.com/cucumber/cucumber-js/blob/main/docs/custom_formatters.md) API and the events published by the `eventBroadcaster`. Could we add those to the public API as well? See: https://github.com/stryker-mutator/stryker-js/blob/03b1f20ed933d3a50b52022cfe363c606c2b16c5/packages/cucumber-runner/src/stryker-formatter.ts#L45-L69

**Comment by davidjgoss** [comment]
> Preferably I would also have an API to load profiles using the same rules as cucumber-js does, so I can mimic the exact behavior of a normal cucumber-js call.

This is a good point. Profiles (in their current form at least) are fundamentally coupled to the CLI so it feels right to keep them on that side of the boundary, but we could still expose a function to load them and generate a partial options object.

**Comment by davidjgoss** [comment]
> While specifying that new public API, we may also consider what recently happened with issue #1489 and think about providing public APIs to have more and better interaction with the filters and the resulting features under test.

I think we could include an option for providing a custom pickle filter when calling the API (in addition to the names, tags etc that drive the built-in filtering).

**Comment by aslakhellesoy** [comment]
The current syntax for [profiles](https://github.com/cucumber/cucumber-js/blob/main/docs/profiles.md) if very command-line-y.

I would ❤️❤️❤️ to be able to specify profiles in a more generic format such as JSON, JavaScript, YAML or environment variables. In JSON it could look like this:

.cucumber.json
```json
{
  "default": {
    "requireModule": ["ts-node/register"],
    "require": ["support/**/*./ts"],
    "worldParameters": {
      "appUrl": "http://localhost:3000/",
    },
    "format": ["progress-bar", "html:./cucumber-report.html"]
  },
  "ci": {
    "requireModule": ["ts-node/register"],
    "require": ["support/**/*./ts"],
    "worldParameters": {
      "appUrl": "http://localhost:3000/",
    },
    "format": ["html:./cucumber-report.html"],
    "publish": true
  }
}
``` 

Or, using JavaScript

.cucumber.js
```javascript
const common = {
  "requireModule": ["ts-node/register"],
  "require": ["support/**/*./ts"],
  "worldParameters": {
    "appUrl": "http://localhost:3000/",
  }
}

module.exports = {
  default: {
    ...common,
    "format": ["progress-bar", "html:./cucumber-report.html"]
  },
  ci: {
    ...common,
    "format": ["html:./cucumber-report.html"],
    "publish": true
  }
}
```

Or even with environment variables (for example loaded with a tool like dotenv):

```env
CUCUMBER_PROFILE_DEFAULT_REQUIREMODULE=ts-node/register
CUCUMBER_PROFILE_DEFAULT_REQUIRE=ts-node/register
CUCUMBER_PROFILE_DEFAULT_WORLDPARAMETERS_APPURL=http://localhost:3000/
CUCUMBER_PROFILE_DEFAULT_FORMAT=progress-bar,html:./cucumber-report.html
CUCUMBER_PROFILE_CI_REQUIREMODULE=ts-node/register
CUCUMBER_PROFILE_CI_REQUIRE=ts-node/register
CUCUMBER_PROFILE_CI_WORLDPARAMETERS_APPURL=http://localhost:3000/
CUCUMBER_PROFILE_CI_FORMAT=progress-bar,html:./cucumber-report.html
CUCUMBER_PROFILE_CI_PUBLISH=true
```

In fact, the [config](https://github.com/cucumber/common/tree/main/config#readme) library does exactly this. We never ended up integrating it in Cucumber-JVM because other stuff got in the way, but maybe we could give it a go with a JavaScript implementation?

**Comment by davidjgoss** [comment]
@aslakhellesoy agree that would be awesome! I'll try and get a POC going for this proposal so we have something a bit more concrete to talk around, and would love to make profiles right as part of it (4.5 years and counting on #751 😄)

**Comment by aurelien-reeves** [comment]
Refs. #1004

**Comment by nicojs** [comment]
> This is a good point. Profiles (in their current form at least) are fundamentally coupled to the CLI so it feels right to keep them on that side of the boundary, but we could still expose a function to load them and generate a partial options object.

Yes, that would be awesome and much appreciated from the point of view of plugin creators.

> I would ❤️❤️❤️ to be able to specify profiles in a more generic format such as JSON, JavaScript, YAML or environment variables. In JSON it could look like this:

That sounds great! And is also exactly the reason I would appreciate an API to load them the same way as cucumberJS does. Loading a single `cucumber.js` file is trivial. Replicating a configuration file loading algorithm, including precedence, file format, etc AND maintaining it is something else entirely 😅.

**Comment by nicojs** [comment]
Q: Would I be able to run `runCucumber` twice in succession _without clearing the [require cache](https://nodejs.org/api/modules.html#modules_require_cache)_? This is important for the mutation testing use case.

We want to load the environment and run the tests multiple times in quick succession while changing a global variable in order to switch the active mutant. 

Right now, we're [using the `cli` private API](https://github.com/stryker-mutator/stryker-js/pull/2970) and we need to clear the step definition files from the `require.cache` between each test run. This is not ideal for CommonJS and won't work at all for esm.

Pseudo code of our use case:

```js
const profiles = await loadProfiles();
const options = {
  ...profiles.default,
  formatter: require.resolve('./our-awesomely-crafted-formatter'),
  some: 'other options we want to override',
}
const cucumber = new Cucumber(options);

// Allow cucumber to load the step definitions once. 
// This is `async`, so support for esm can be added without a breaking change
await cucumber.initialize();

// Initial test run ("dry run"), without mutants active
await cucumber.run();

collectMutantCoveragePerTestFromFormatter();

// Start mutation testing:

global.activeMutant = 1;
await cucumber.run({ features: ['features/a.feature:24']);
collectResultsFromFormatterToDetermineKilledOrSurvivedMutant()

global.activeMutant = 2;
await cucumber.run({ features: ['features/b.feature:24:25:26', 'features/c.feature:12']);
collectResultsFromFormatterToDetermineKilledOrSurvivedMutant()

// etc
```

**Comment by nicojs** [comment]
That works for us 👍

**Comment by jan-molak** [comment]
Works for us too 👍🏻

**Comment by rkrisztian** [comment]
I believe something like this would be very useful as an alternative for the mess we have today with testing tool integrations (Jest, Cypress), for example:

* `jest-cucumber` is not as convenient as `jest-cucumber-fusion`
* there's also `cucumber-jest`...
* `jest-cucumber` does not support Cucumber JSON reporting (https://github.com/bencompton/jest-cucumber/issues/27)
* `cypress-cucumber-preprocessor` generates multiple Cucumber JSON reports with no offical support for aggregation (https://github.com/TheBrainFamily/cypress-cucumber-preprocessor/issues/423)

I would rather see some minimal glue code between Jest/Karma/Cypress/etc. and cucumber-js so I don't have to suffer for all those missing features I need to use.

**Comment by mattwynne** [comment]
Great suggestion @davidjgoss 👍 

This separation of concerns between the command-line user interface and the "business logic" of parsing and executing scenarios as tests reminds me of the [hexagonal architecture](https://alistair.cockburn.us/hexagonal-architecture/) pattern.

In cucumber-ruby we actually split the core domain logic (or "inner hexagon") into a separate gem package, as we were rebuilding it from scratch in a "clean room". I realise that's not the context here, but it might be worth drawing some inspiration from, or feeding back innovations from this design into the Ruby API. There's [an example](https://github.com/cucumber/cucumber-ruby-core#example) in the cucumber-ruby-core gem's README of how to use that API.

**Comment by davidjgoss** [comment]
Okay, here's a first pass at the API signature for the "run" bit. It's heavily based on the `IConfiguration` object we have internally (so shouldn't cause too much refactoring lower down) but just a little less "flat":

```typescript
export interface IRunCucumberOptions {
  cwd: string
  features: {
    defaultDialect?: string
    paths: string[]
  }
  filters: {
    names?: string[]
    tagExpression?: string
  }
  support:
    | {
        transpileWith?: string[]
        paths: string[]
      }
    | ISupportCodeLibrary
  runtime: {
    dryRun?: boolean
    failFast?: boolean
    filterStacktraces?: boolean
    parallel?: {
      count: number
    }
    retry?: {
      count: number
      tagExpression?: string
    }
    strict: boolean
    worldParameters?: any
  }
  formats: {
    stdout: string
    files: Record<string, string>
    options: IParsedArgvFormatOptions
  }
}

export interface IRunResult {
  success: boolean
  support: ISupportCodeLibrary
}

export async function runCucumber(
  options: IRunCucumberOptions
): Promise<IRunResult> {
  // do stuff
}
```

And a very contrived example usage:

```typescript
const result = await runCucumber({
  cwd: process.cwd(),
  features: {
    paths: ['features/**/*.feature'],
  },
  filters: {
    names: ['Acme'],
    tagExpression: '@interesting',
  },
  support: {
    transpileWith: ['ts-node'],
    paths: ['features/support/**/*.ts'],
  },
  runtime: {
    failFast: true,
    retry: {
      count: 1,
      tagExpression: '@flaky',
    },
    strict: true,
    worldParameters: {
      foo: 'bar',
    },
  },
  formats: {
    stdout: '@cucumber/pretty-formatter',
    files: {
      'report.html': 'html',
      'TEST-cucumber.xml': 'junit',
    },
    options: {
      printAttachments: false,
    },
  },
})
```

Feedback welcome! Note this doesn't cover the profile/config loading stuff which would be another function.

**Comment by nicojs** [comment]
I think this looks good. Question: How would I configure a custom formatter?

**Comment by davidjgoss** [comment]
@nicojs sorta like on the CLI

```typescript
formats: {
    files: {
      'report.html': './my/fancy-reporter.js',
      'other-report.html': '@me/reporter-package',
    }
  },
```

**Comment by aslakhellesoy** [comment]
Great to see progress on this @davidjgoss!

I don't want to slow down the progress on this, but at the same time I want to make sure we adopt a format that can work for other Cucumber implementations too.

Eventually a JSON schema, but while we discuss it I think TypeScript types are easier for us humans to parse.

I suggest we create a new issue about the proposed format in the `cucumber/common` monorepo and invite the core team to discuss there.

**Comment by davidjgoss** [comment]
@aslakhellesoy will do.

What would you think about the functional API not being tied to the common options structure? Like we'd map from that to the `runCucumber` options. It adds maybe a little complexity but appeals because of things like having a `support` block that's either parameters to load, or a previously loaded support code library. Could be the same for features+pickles too.

**Comment by aslakhellesoy** [comment]
> What would you think about the programmatic API not being tied to the common options structure?

I think that's fine, as long as we provide a function that converts from the options file contents to the data structure `runCucumber` wants.

**Comment by nicojs** [comment]
> Eventually a JSON schema, but while we discuss it I think TypeScript types are easier for us humans to parse.

Why do we need to choose? We're using JSON schema in StrykerJS to generate typescript using [json-schema-to-typescript](https://www.npmjs.com/package/json-schema-to-typescript). We're not committing the TS output files to source control, instead we generate them on the fly using a `prebuild` step.

JSON schemas are still somewhat readable for humans IMO. We've already had PR's on the Stryker repo and people seem to know what to do 🤷‍♀️

> sorta like on the CLI
> 
> ```ts
> formats: {
>     files: {
>       'report.html': './my/fancy-reporter.js',
>       'other-report.html': '@me/reporter-package',
>     }
>   },
> ```

How would this work for a reporter that doesn't need an output file name? Like so:

```ts
formats: {
  files: {
    '': require.resolve('./my-awesome-stryker-formatter')
  }
}
```

**Comment by aslakhellesoy** [comment]
> Why do we need to choose?

I think we should use a JSON Schema as a single source of truth for the structure of the configuration. -And then generate TypeScript/Java/Whatever code from that schema.

But JSON Schema is a bit hard to read for humans, so while we are *discussing* the schema in a GitHub issue in `cucumber/common` I was suggesting TypeScript to facilitate the discussion.

See what I mean?

**Comment by aslakhellesoy** [comment]
> JSON schemas are still somewhat readable for humans IMO

Not to me :-) Too verbose.

**Comment by davidjgoss** [comment]
> How would this work for a reporter that doesn't need an output file name?

```typescript
formats: {
    stdout: './my-awesome-stryker-formatter',
    files: {
      'report.html': './my/fancy-reporter.js',
      'other-report.html': '@me/reporter-package',
    }
  },
```

**Comment by shawnbot** [comment]
This sounds great to me. Over in #1724 I was asking if there's a way to modify Cucumber step results so that I can integrate with a test runner that needs the step configuration upfront (which I was hoping to build up from step definitions), runs the whole test suite (scenario) async, then returns results for each step. Typically this would be done in a hook, but none of the Cucumber hooks runs at the right time or gives me (public) access to the step results so that I can update them before they're passed to the formatter.

I could probably do this work in a custom formatter, but that involves (AFAICT) having to share state between the custom world objects (that build up test suites) and the formatter, which feels like a gross hack. Wrapping the test runner programmatically would at least make it easier and a bit more graceful. Without a hooks-like API for `runCucumber()`, I would expect to be able to do something like this:

```js
import { runCucumber, formatResults } from '@cucumber/cucumber'
import { cucumberDefaults, transformTestResults } from '@sfgov/ghost-cucumber'

const results = await runCucumber({
  ...cucumberDefaults,
  formats: { /* no formatting? */ }
})

const modifiedResults = await transformTestResults(results)

formatResults(modifiedResults, {
  // use the default formatter if none is provided?
  file: '/dev/stdout'
})
  .then(({ status }) => process.exit(status)) // does the formatter report status, or...?
```

**Comment by OhadR** [comment]
note that i have spent HOURS on this, trying to run cucumber by its Cli. After struggling, I have noticed that it can be invoked only once, and in the second run it does not find the step-definitions (because of `SupportCodeLibraryBuilder` - nodejs cache the step-definitions so in the 2nd run it does not "require" them again, in `getSupportCodeLibrary()`)

**Comment by davidjgoss** [comment]
@OhadR we're going to be catering for this as part of the new `runCucumber` API which is planned for 8.0.0- see https://github.com/cucumber/cucumber-js/issues/1894.

In the meantime there are workarounds with removing things from the require cache yourself https://github.com/cucumber/cucumber-js/issues/1894

**Comment by binarymist** [comment]
> In the meantime there are workarounds with removing things from the require cache yourself
Or Just create a [process for each cuc cli instance](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.cuc.js#L32) as we did about 3 years ago.

**Comment by binarymist** [comment]
> Currently we don't have a good way to programmatically run cucumber-js

Hmm, we've been running it fine for four years.

> Projects that use cucumber-js as part of a framework (e.g. Serenity, Stryker)

You could add [PurpleTeam](https://purpleteam-labs.com) to this list.


Currently with the move from:

* @cucumber/cucumber 7.2.1 to 8.0.0-rc.2
* @cucumber/gherkin-streams 2.0.2 to 4.0.0

We have multiple ([app-scanner](https://github.com/purpleteam-labs/purpleteam-app-scanner), [tls-scanner](https://github.com/purpleteam-labs/purpleteam-tls-scanner), server-scanner soon) _Testers_, architecture diagram [here](https://purpleteam-labs.com/doc/cloud/). For all of our _Testers_ our [CLI](https://github.com/purpleteam-labs/purpleteam)'s `test` command still works, but our `testplan` is broken for all _Testers_.

The following example is of our app-tester:

## `testplan`

1. For the [`testplan`](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.js#L136-L137) command we...
2. [`createCucumberArgs`](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.js#L153-L196)
3. We pass the [new `CucCli` instance](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.js#L138) to [`getActiveFeatureFileUris`](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.js#L199)
4. Which calls the legacy `getConfiguration` method of the cucumber CLI instance which no longer exists

What should we now use to replace this removed functionality?



## `test`

In the app-tester we `spawn` multiple cucumber CLI instances based on the number of [_TestSessions_](https://purpleteam-labs.com/doc/definitions/) passed from the CLI to the _orchestrator_ to the _Testers_.

1. We create the cucumber args like [this](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.js#L153-L196)
2. Then we `spawn` our [bin file](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/bin/purpleteamParallelCucumber) like [this](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/6d28344dfea0519ac4d961578ba07de785ed252c/src/api/app/models/app.cuc.js#L32-L61)
3. Which instantiates and runs the cucumber CLI like (this](https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/main/src/scripts/runCuc.js#L64-L72)

Will this still work with the proposed changes, if not what will we need to do to refactor?

@davidjgoss does your `IRunCucumberOptions` example [here](https://github.com/cucumber/cucumber-js/issues/1711#issuecomment-876428621) include: path to step definitions?

What would we need to change to move to the new `runCucumber`?

**Comment by davidjgoss** [comment]
Thanks for the detailed info @binarymist, that will be very useful. For now I’d say sit tight until the next RC as there is still more to do on this piece eg configuration loading. It’ll be fully documented when ready.

**Comment by binarymist** [comment]
Hi @davidjgoss 

> @aurelien-reeves mentioned in Slack:

>> A lot may have changed, but actually you should be able to find the same things than before, they may have just moved.

> What should we now use to replace this removed functionality?

This is our imediate problem... Any feedback on this?

Thanks.

**Comment by nicojs** [comment]
@davidjgoss I might have a use case to be able to load the support code first, activate the mutant and then run the tests. I think we would be able to implement this feature through one of the `on('envelope')` events in our custom `Formatter`, but maybe it is easy for you to add this on its own?

```js
const support = await loadSupport({
    transpileWith: ['ts-node'],
    paths: ['features/support/**/*.ts'],
});
globalThis.__stryker__.activeMutant = "1";
const result = await runCucumber({ support });
```

See https://github.com/stryker-mutator/stryker-js/issues/3442 for a bunch of details 😉

**Comment by davidjgoss** [comment]
@nicojs that should be possible although it would also require the "sources" part of the configuration, since the resolution of support code takes the feature paths into account. But as a consumer I think you would just need to load the whole configuration object anyway. So like:

```ts
import {loadConfiguration, loadSupport, runCucumber} from '@cucumber/cucumber`

const configuration = await loadConfiguration()
const support = await loadSupport(configuration)
const result = await runCucumber({...configuration, support})
```

Does that make sense?

**Comment by nicojs** [comment]
That would be sweet 😌. Let me know when we can do tests with this setup.

**Comment by davidjgoss** [comment]
Just calling out that as we'll have `loadSupport()` to load the support code separately to actually doing a test run, we might want something similar like `loadSources()` to do the Gherkin loading and processing as a standalone thing. A good use case is the example from @binarymist where a test plan is calculated and reported without doing a run.

**Comment by nicojs** [comment]
That would also help mutation testing performance a lot! 😍

**Comment by davidjgoss** [comment]
@jan-molak yeah i didn't describe it that well but I think my `loadSources` and your `loadSpecs` are kinda the same idea - `loadSources` would be giving you the features and pickles, either pre-filtered or with the easy means to filter (based on tags, names, urls, lines).

Cucumber has the concept of a "test case" but these are the composition of filtered pickles with support code including hooks etc which I don't think is right for the use case you describe - you're just interested in which scenarios are going to get run regardless of the state of support code, right?

**Comment by jan-molak** [comment]
Hey @davidjgoss!

> you're just interested in which scenarios are going to get run regardless of the state of support code, right?

Let me think aloud :) Consider this example. Let's say we have a couple of feature files like the ones below:

```gherkin

# features/first.feature

Feature: First feature                            # 1
                                                  # 2
  @current                                        # 3
  Scenario: First scenario                        # 4
                                                  # 5
  Scenario: Second scenario                       # 6

# features/second.feature

Feature: Second feature

  Scenario: Some other scenario
```

When I ask Cucumber to run:
```
await loadSpecs({
  specs: [`features/first.feature`]
  tags: []
})
```
I'd like it to tell me that it will run:
```
[
  `features/second.first:4`
  `features/second.first:6`
]
```

When I ask it to run:
```
await loadSpecs({
  specs: [`features/first.feature:4`]
  tags: []
})
```
I'd like it to tell me that it will run:
```
[
  `features/second.first:4`
]
```

When I use tags (or any other filter):
```
await loadSpecs({
  specs: [`features/*.feature`]
  tags: [`@current`]
})
```
I'd like it to tell me that it will run:
```
[
  `features/second.first:4`
]
```

So, given a configuration that includes paths to feature files, filters, tags, etc, I'd like to receive a list of scenario locations that Cucumber is going to execute when I invoke `run`.

When the list is empty, I know I don't need to start the browser.
When it's non-empty, I could use it to "load-balance" scenarios between browsers.

Does it make sense?

**Comment by davidjgoss** [comment]
@jan-molak it does, thanks! I can't promise it'll look exactly like that, but it'll be possible.

**Comment by davidjgoss** [comment]
Quick follow-up on the `loadSources`/`loadPickles`/`loadSpecs` (or whatever we call it) function which I'm going to work on today. There are two possible use cases I can see:

- "Calculate a test plan" - load the features, parse to pickles, filter the pickles (based on paths, names, tags) and make it easy to tell what will be run without actually doing a run. You _could_ achieve this by using `dryRun` and picking through the messages but that's pretty unwieldy and also loads the support code unnecessarily.
- "Avoid reloading the features" - load the features, parse to pickles, and return those messages (`source`, `gherkinDocument`, `pickle`) so I can pass them in to as many `runCucumber` calls as I want to make without having them loaded and parsed again each time. I'd note this could be a very large amount of data, as it would be the original source, AST and compiled pickles for every feature loaded.

My thought for 8.0.0 is to focus on the first one but ensure the API is flexible enough to add the second one later in a non-breaking way. Thoughts?

**Comment by aurelien-reeves** [comment]
Good plan 👌

My though is more related to v8.0.0: did we released 8.0.0-rc.1 a little bit early? 🤔

**Comment by davidjgoss** [comment]
> did we released 8.0.0-rc.1 a little bit early?

Yes, I think so in retrospect.

**Comment by davidjgoss** [comment]
[8.0.0-rc.3 has been released](https://github.com/cucumber/cucumber-js/releases/tag/v8.0.0-rc.3) including this new API, [there are some docs here](https://github.com/cucumber/cucumber-js/blob/main/docs/javascript_api.md) - a little thin at the moment but will continue to expand them.

If anyone has any early feedback that'd be fantastic (cc @jan-molak @nicojs @binarymist).

**Comment by MarufSharifi** [comment]
> 

Hey thanks for doing this great job, I need to use from @cucumber/cucumber: 8.0.0, when it will be available to use?

**Comment by davidjgoss** [comment]
@MarufSharifi within the next week or two, but you can try `8.0.0-rc.3` in the meantime if that helps.

**Comment by binarymist** [comment]
Thanks @davidjgoss .

https://github.com/cucumber/cucumber-js/releases/tag/v8.0.0-rc.3 links to https://github.com/cucumber/cucumber-js/blob/v8.0.0-rc.3/docs/cli.md#printing-attachments-details which doesn't appear to exist.

---

I'm assuming https://github.com/cucumber/cucumber-js/blob/main/docs/javascript_api.md#sources-example is what I'm looking for to recreate our testplan functionality?
I'm not quite sure how to use this, with the following in source code:

`import { loadConfiguration, loadSources } from '@cucumber/cucumber/api'` 

and the following in the package.json:

`"@cucumber/cucumber": "^8.0.0-rc.3"`

we get:

`Unable to resolve path to module '@cucumber/cucumber/api'.`

Then if we try the following in the package.json:
```
    "@cucumber/cucumber": "^8.0.0-rc.2",
    "@cucumber/cucumber/api": "^8.0.0-rc.3",
```
We get :
```
npm ERR! Invalid package name "@cucumber/cucumber/api": name can only contain URL-friendly characters
```

**Comment by binarymist** [comment]
Thanks @davidjgoss .

https://github.com/cucumber/cucumber-js/releases/tag/v8.0.0-rc.3 links to https://github.com/cucumber/cucumber-js/blob/v8.0.0-rc.3/docs/cli.md#printing-attachments-details which doesn't appear to exist.

---

I'm assuming https://github.com/cucumber/cucumber-js/blob/main/docs/javascript_api.md#sources-example is what I'm looking for to recreate our testplan functionality?

`loadSources(runConfiguration.sources)` returns `undefined`

Can you please explain how this can replace our existing `testPlan` functionality?

Thanks.

**Comment by aurelien-reeves** [comment]
> Thanks @davidjgoss .
> 
> https://github.com/cucumber/cucumber-js/releases/tag/v8.0.0-rc.3 links to https://github.com/cucumber/cucumber-js/blob/v8.0.0-rc.3/docs/cli.md#printing-attachments-details which doesn't appear to exist.
> 

Fixed, thanks for the notification

**Comment by aurelien-reeves** [comment]
> 



> I'm assuming https://github.com/cucumber/cucumber-js/blob/main/docs/javascript_api.md#sources-example is what I'm looking for to recreate our testplan functionality?
> 
> `runConfiguration` doesn't seem to contain anything of value.
> 
> `loadSources(runConfiguration.sources)` returns `undefined`
> 
> Can you please explain how this can replace our existing `testPlan` functionality, maybe even provide some tips to help with the guess work?
> 
> Thanks.

loadConfiguration returns a `runConfiguration` and a `useConfiguration`. `useConfiguration` is the resolution of the configuration that the API has computed from the configuration file, environment, and other given parameters.

In your case, how does it looks like? If it does not looks as expected, give the proper argument to `loadConfiguration` method.

Refs to the documentation of the API:
- IResolvedConfiguration: https://github.com/cucumber/cucumber-js/blob/main/docs/api/cucumber.iresolvedconfiguration.md
- LoadConfiguration method: https://github.com/cucumber/cucumber-js/blob/main/docs/api/cucumber.loadconfiguration.md

If you are still not able to make it work, please provide more info to us like your piece of code, what is expected, and what is the actual result

**Comment by davidjgoss** [comment]
@binarymist so for the test plan case specifically you shouldn't need to use the `Cli` class at all. `loadSources` from the new API should, I think, give you what you need.

The `loadConfiguration` method is designed to source configuration from the standard cucumber [configuration file](https://github.com/cucumber/cucumber-js/blob/main/docs/configuration.md#files) that a user might have in their project, apply some defaults where nothing is specified, and return it in a shape that works for the other API functions. For your case it seems configuration is managed in its own way and you already have it at runtime, so you could use the `provided` option to include that directly. This would be similar to what you do with `createCucumberArgs`, except it's [in object form](https://github.com/cucumber/cucumber-js/blob/main/docs/configuration.md#options) rather than an argv-style string, something like:

```js
const { runConfiguration } = await loadConfiguration({
  provided: {
    paths: [`${this.#cucumber.features}/${sessionProps.sUtType}`],
    require: [`${this.#cucumber.steps}/${sessionProps.sUtType}`],
    format: [`message:${this.#results.dir}result_appScannerId-${sessionProps.testSession ? sessionProps.testSession.id : 'noSessionPropsAvailable'}_${this.#strings.NowAsFileName('-')}.NDJSON`],
    // etc
  }
})
```

`runConfiguration.sources` should then be what you need to give to `loadSources`, which returns an object with a `plan` property, which is a (pre-filtered by paths, tags, names etc) array of objects each representing a pickle (test case) to be run.

Both functions have an optional `environment` argument which you mentioned - you don't need to provide this unless you need any of the values to be different from the defaults, which are:

```js
{
  cwd: process.cwd(), // files would be loaded relative to this dir
  stdout: process.stdout,
  stderr: process.stderr,
  env: process.env,
}
```

I hope this helps. I appreciate the docs are a little thin right now but they'll improve quickly and feedback like this is helpful.

**Comment by garfieldnate** [comment]
I might be too late to the party to give input, but the API still seems fairly similar to running cucumber from the CLI, and I'm having some trouble understanding how to apply this for my use case.

I want to write a test file that contains all of the steps, then run those steps on one or more associated feature-files. I'm using TS, and I want to be able to compile the test and then run it repeatedly as I develop the feature file without having to re-do the transpilation step repeatedly, which slows down the development loop.

It's already been mentioned that I can use `loadConfiguration({provided: {...}})` to define the config inline. I'm unsure of how to do the same for `loadSupport`; I need to create an `ISupportCodeLibrary` from the steps and world in the current file, but there's no documentation on how to construct this.

**Comment by binarymist** [comment]
Thanks @davidjgoss.

That seems to have fixed the `testPlan` functionality, but it looks like the update to rc.3 has broken our `test` functionaly which of course is a show stopper.

Have you seen this error before:

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /usr/src/app/src/steps/BrowserApp/app_scan_steps.js from /usr/src/app/node_modules/@cucumber/cucumber/lib/api/support.js not supported.
 Instead change the require of app_scan_steps.js in /usr/src/app/node_modules/@cucumber/cucumber/lib/api/support.js to a dynamic import() which is available in all CommonJS modules.
     at /usr/src/app/node_modules/@cucumber/cucumber/lib/api/support.js:14:32
     at Array.map (<anonymous>)
     at getSupportCodeLibrary (/usr/src/app/node_modules/@cucumber/cucumber/lib/api/support.js:14:18)
     at runCucumber (/usr/src/app/node_modules/@cucumber/cucumber/lib/api/run_cucumber.js:33:53)
     at async Cli.run (/usr/src/app/node_modules/@cucumber/cucumber/lib/cli/index.js:45:29)
     at async run (file:///usr/src/app/src/scripts/runCuc.js:75:14) {
   code: 'ERR_REQUIRE_ESM'
 }

```

**Comment by binarymist** [comment]
There doesn't appear to be a lib/api/support.js in the root of the [cucumber repo](https://github.com/cucumber/cucumber-js/tree/release/v8.0.0-rc.3), but I found a src/api/support.ts which appears to be new in rc.3 and it appears to be `require`ing our ESM, so as expected the error makes perfect sense.

So it seems that cucumber in rc.2 supports ESM but rc.3 no longer does?

**Comment by davidjgoss** [comment]
@binarymist it’s still there but uses a specfic option for import vs require now - previously it was inferred but the heuristics were too fuzzy so we changed it to be explicit:

- https://github.com/cucumber/cucumber-js/pull/1931
- https://github.com/cucumber/cucumber-js/blob/main/docs/esm.md

**Comment by davidjgoss** [comment]
(We should add better messaging for that, thanks for raising.)

**Comment by davidjgoss** [comment]
@garfieldnate just to make sure I understand, you want to:

- Load (including transpile) the support code _once_
- Use that support code on multiple test runs, triggered when you change a feature file

If so, I think you could implement this now. You can't create the support code library inline, but you can load it once and use it multiple times. Something like:

```js
import {loadConfiguration, loadSupport, runCucumber} from '@cucumber/cucumber/api'

const {runConfiguration} = await loadConfiguration()
const support = await loadSupport(runConfiguration)
const merged = {...runConfiguration, support}

// do this bit as many times as you want e.g. in a loop or callback
await runCucumber(merged)
```

**Comment by binarymist** [comment]
Thanks @davidjgoss.

Initial testing has the import working. It seems that the colour characters (green ticks, red crosses, dashes, etc) that are usually written are no longer colour. We haven't changed the `'--format-options', '{"colorsEnabled": true}',`

**Comment by aurelien-reeves** [comment]
That seems to be an issue.
Indeed there have been recent changes under the hood, but not related to the API

Do you think you could submit a new issue for that, with a minimal reproducible example?

**Comment by davidjgoss** [comment]
Worth noting we switched from using `colors` to `chalk` in a hurry for well-known reasons, could be related to that.

**Comment by binarymist** [comment]
> Do you think you could submit a new issue for that, with a minimal reproducible example?

Currently we're going through the release phase with no colours (as a known issue). We're unable to keep PurpleTeam users waiting any longer. Will try and circle back to this. I'm not sure how minimal a seperate reproducable example would be. My guess is that everyone else expecting colours on this version will also be facing the same problem. It's just running the CLI with `'--format-options', '{"colorsEnabled": true}'`

**Comment by aurelien-reeves** [comment]
> It's just running the CLI with `'--format-options', '{"colorsEnabled": true}'`
> Nothing else special, we've been running the CLI this way for years.

I already tried that, but I cannot reproduce your issue

![image](https://user-images.githubusercontent.com/24386780/165466125-814815af-a593-403e-b216-f4d268399fc7.png)


How do you pass the format options to cucumber? Through the CLI directly? Using a config file?

I am puzzled with the format you are using. Why such format? I mean: `'--format-options', '{"colorsEnabled": true}'`: how do you pass those parameters to cucumber?

First of all, that option is true per default, so you should be able to not specifying that option.
Also, why that coma between format-options and the options? And the quotes? I would suggest trying this instead: ` --format-options '{"colorsEnabled": false}'`

**Comment by davidjgoss** [comment]
@aurelien-reeves I think that's from running the `Cli` programmatically where you provide argv as `string[]`

I think I might see the/an issue here, going to raise something new though @binarymist as it's not API related per se.
